io/src/main/scala/fs2/io/file/Watcher.scala

```diff
@@ -288,45 +302,50 @@ object Watcher {
-              )
-            )
-            .getOrElse(Nil)
+          val regs = registrations.filter(_.key == key)
+          val filteredEvents = events.filter { e =>
+            regs.foldLeft(false) { (acc, reg) =>
+              acc || {
+                val singleFileMatch = reg.singleFile && Event.pathOf(e) == Some(reg.path)
+                val dirMatch =
+                  !reg.singleFile && !(e.isInstanceOf[Event.Created] && reg.suppressCreated)
+                singleFileMatch || dirMatch
+              }
+            }
+          }
           val recurse: Stream[F, Event] =
-            if (reg.map(_.recurse).getOrElse(false)) {
+            if (regs.map(_.recurse).foldLeft(false)(_ || _)) {
               val created = events.collect { case Event.Created(p, _) => p }
               def watchIfDirectory(p: Path): F[(F[Unit], List[Event])] =
                 F.blocking(Files.isDirectory(p))
                   .ifM(
-                    watch(p, Seq(EventType.Created), reg.map(_.modifiers).getOrElse(Nil)).flatMap {
-                      cancel =>
-                        val events: F[List[Event]] = F.blocking {
-                          var dirs: List[Path] = Nil
-                          Files.list(p).forEach(d => dirs = d :: dirs)
-                          dirs.map(Event.Created(_, 1))
-                        }
-                        events.map(cancel -> _)
+                    watch(
+                      p,
+                      Seq(EventType.Created),
+                      regs.headOption.map(_.modifiers).getOrElse(Nil)
+                    ).flatMap { cancel =>
+                      val events: F[List[Event]] = F.blocking {
+                        var dirs: List[Path] = Nil
+                        Files.list(p).forEach(d => dirs = d :: dirs)
+                        dirs.map(Event.Created(_, 1))
+                      }
+                      events.map(cancel -> _)
                     },
                     F.pure(F.unit -> List.empty[Event])
                   )
               val subregisters: F[List[Event]] =
                 created.traverse(watchIfDirectory).flatMap { x =>
                   val (cancels, events) = x.separate
                   val cancelAll: F[Unit] = cancels.sequence.void
-                  val updateRegistration: F[Unit] = this.registrations
-                    .update(m =>
-                      m.get(key)
-                        .map(r => m.updated(key, r.copy(cleanup = r.cleanup >> cancelAll)))
-                        .getOrElse(m)
-                    )
-                    .void
+                  val updateRegistration: F[Unit] = this.registrations.update { m =>
+                    val idx = m.indexWhere(_.key == key) // TODO indexOf(_)
+                    if (idx >= 0) {
+                      val existing = m(idx)
+                      m.updated(idx, existing.copy(cleanup = existing.cleanup >> cancelAll))
+                    } else m
+                  }.void
```

Comment:
`cancelAll` can be moved inside the `if` statement to avoid an allocation when it is not necessary to evaluate that branch.